// @braze/web-sdk@5.2.0/src/managers/network-manager.js
import qt from "../models/backend-errors.js";
import ue from "../models/braze-event.js";
import {
  convertMsToSeconds as h,
  convertSecondsToMs as Xt,
} from "../util/date-utils.js";
import s from "../common/event-logger.js";
import { isArray as j, isEqual as ii } from "../util/code-utils.js";
import { logger as r, EventTypes as i } from "../../shared-lib/index.js";
import { STORAGE_KEYS as o } from "./storage-manager.js";
import S from "../util/request-header-utils.js";
import { LAST_REQUEST_TO_ENDPOINT_MS_AGO_DEFAULT as Kt } from "../common/constants.js";
import { getAlias as Ht } from "./utils.js";
export default class Rt {
  constructor(t, e, s, i, r, n, o, h, a, u, l, c) {
    (this.on = t),
      (this.u = e),
      (this.Go = s),
      (this.wt = i),
      (this.di = r),
      (this.yt = n),
      (this.rn = o),
      (this.Ko = h),
      (this.Wo = a),
      (this.Vo = u),
      (this.appVersion = l),
      (this.Xh = c),
      (this.Hh = (t) => (null == t ? "" : `${t} `)),
      (this.on = t),
      (this.u = e),
      (this.Go = s),
      (this.wt = i),
      (this.di = r),
      (this.yt = n),
      (this.rn = o),
      (this.Ko = h),
      (this.Wo = a),
      (this.Vo = u),
      (this.appVersion = l),
      (this.Xh = c),
      (this.Yh = ["npm"]);
  }
  Ps(t, e = !1, s = !1) {
    const i = this.on.ce(!s),
      r = i.Mr(),
      n = this.u.j(o.C.Ga);
    ii(n, r) || (t.device = r),
      (t.api_key = this.rn),
      (t.time = h(new Date().valueOf(), !0));
    const a = this.u.j(o.C.Jh) || [],
      u = this.u.j(o.C.Qh) || "";
    if (
      (this.Yh.length > 0 &&
        (!ii(a, this.Yh) || u !== this.di.ki()) &&
        (t.sdk_metadata = this.Yh),
      (t.sdk_version = this.Wo),
      this.Vo && (t.sdk_flavor = this.Vo),
      (t.app_version = this.appVersion),
      (t.app_version_code = this.Xh),
      (t.device_id = i.id),
      e)
    ) {
      const e = this.wt.getUserId();
      null != e && (t.user_id = e);
    }
    if (!t.user_id && !this.Go.Fh()) {
      const e = Ht(this.u);
      e && (t.alias = e);
    }
    return t;
  }
  Ws(t, e, n) {
    const o = e.auth_error,
      h = e.error;
    if (!o && !h) return !0;
    if (o) {
      let e;
      this.Go.Bh();
      const s = { errorCode: o.error_code };
      for (const t of n)
        j(t) && "X-Braze-Auth-Signature" === t[0] && (s.signature = t[1]);
      t.respond_with && t.respond_with.user_id
        ? (s.userId = t.respond_with.user_id)
        : t.user_id && (s.userId = t.user_id);
      const i = o.reason;
      return (
        i
          ? ((s.reason = i), (e = `due to ${i}`))
          : (e = `with error code ${o.error_code}.`),
        this.Go.Fh() ||
          (e +=
            ' Please use the "enableSdkAuthentication" initialization option to enable authentication.'),
        r.error(`SDK Authentication failed ${e}`),
        this.Zh(t.events || [], t.attributes || []),
        this.Go.qh(s),
        !1
      );
    }
    if (h) {
      let n,
        o = h;
      switch (o) {
        case qt.Ra:
          return (
            (n = "Received successful response with empty body."),
            s.q(i.Ms, { e: n }),
            r.info(n),
            !1
          );
        case qt.Ba:
          return (
            (n = "Received successful response with invalid JSON"),
            s.q(i.Ms, { e: n + ": " + e.response }),
            r.info(n),
            !1
          );
        case qt.Xa:
          o = `The API key "${t.api_key}" is invalid for the baseUrl ${this.Ko}`;
          break;
        case qt.Na:
          o =
            "Sorry, we are not currently accepting your requests. If you think this is in error, please contact us.";
          break;
        case qt.$a:
          o = "No device identifier. Please contact support@braze.com";
      }
      r.error("Backend error: " + o);
    }
    return !1;
  }
  Ea(t, e, s, i) {
    return !!((t && 0 !== t.length) || (e && 0 !== e.length) || s || i);
  }
  Ka(t, e, s, i, r = !1) {
    const n = [],
      o = (t) => t || "",
      h = o(this.wt.getUserId());
    let a = this.Zi(t, e);
    const u = [],
      l = [];
    let c,
      d = null;
    if (s.length > 0) {
      const t = [];
      for (const e of s) {
        if (((c = e.Mr()), this.Go.Fh())) {
          if (h && !c.user_id) {
            d || (d = {}), d.events || (d.events = []), d.events.push(c);
            continue;
          }
          if (o(c.user_id) !== h) {
            l.push(c);
            continue;
          }
        }
        t.push(c);
      }
      t.length > 0 && (a.events = t);
    }
    if (i.length > 0) {
      const t = [];
      for (const e of i)
        e && (this.Go.Fh() && o(e.user_id) !== h ? u.push(e) : t.push(e));
      t.length > 0 && (a.attributes = t);
    }
    if ((this.Zh(l, u), (a = this.Ps(a, !0, r)), d)) {
      d = this.Ps(d, !1, r);
      const t = { requestData: d, headers: this._s(d, S.Gs.La) };
      n.push(t);
    }
    if (a && !this.Ea(a.events, a.attributes, t, e)) return d ? n : null;
    const f = { requestData: a, headers: this._s(a, S.Gs.La) };
    return n.push(f), n;
  }
  Zh(t, e) {
    if (t) {
      const e = [];
      for (const s of t) {
        const t = ue.fromJson(s);
        (t.time = Xt(t.time)), e.push(t);
      }
      this.u.zo(e);
    }
    if (e) for (const t of e) this.u.Pa(t);
  }
  Zs(t, e) {
    let s = "HTTP error ";
    null != t && (s += t + " "), (s += e), r.error(s);
  }
  qr(t) {
    return s.q(i.Ua, { n: t });
  }
  Zi(t, e, s) {
    const i = {};
    t && (i.feed = !0), e && (i.triggers = !0);
    const r = null != s ? s : this.wt.getUserId();
    if ((r && (i.user_id = r), !i.user_id && !this.Go.Fh())) {
      const t = Ht(this.u);
      t && (i.alias = t);
    }
    return (i.config = { config_time: this.yt.ni() }), { respond_with: i };
  }
  Ha(t) {
    const e = new Date().valueOf();
    let s = Kt.toString();
    const i = S.Wa(this.u, t);
    if (-1 !== i) {
      s = (e - i).toString();
    }
    return s;
  }
  _s(t, e, s = !1) {
    const i = [["X-Braze-Api-Key", this.rn]],
      r = this.Ha(e);
    i.push(["X-Braze-Last-Req-Ms-Ago", r]);
    const n = S.Ya(this.u, e).toString();
    i.push(["X-Braze-Req-Attempt", n]);
    let h = !1;
    if (
      (null != t.respond_with &&
        t.respond_with.triggers &&
        (i.push(["X-Braze-TriggersRequest", "true"]), (h = !0)),
      null != t.respond_with &&
        t.respond_with.feed &&
        (i.push(["X-Braze-FeedRequest", "true"]), (h = !0)),
      e === S.Gs.Bs)
    ) {
      i.push(["X-Braze-ContentCardsRequest", "true"]);
      let t = this.u.j(o.C.As);
      (t && s) || ((t = 0), this.u.k(o.C.As, t)),
        i.push(["BRAZE-SYNC-RETRY-COUNT", t.toString()]),
        (h = !0);
    }
    if (
      (e === S.Gs.qi &&
        (i.push(["X-Braze-FeatureFlagsRequest", "true"]), (h = !0)),
      h && i.push(["X-Braze-DataRequest", "true"]),
      this.Go.Fh())
    ) {
      const t = this.Go.kh();
      null != t && i.push(["X-Braze-Auth-Signature", t]);
    }
    return i;
  }
  Qa(t, e, s, i) {
    window.setTimeout(() => {
      r.info(`Retrying rate limited ${this.Hh(i)}SDK request.`),
        this.Hs(e, s, i);
    }, t);
  }
  Hs(t, e, s) {
    const i = this.Va();
    if (!i.Za)
      return (
        r.info(
          `${this.Hh(
            s,
          )}SDK request being rate limited. Request will be retried in ${Math.trunc(
            i.au / 1e3,
          )} seconds.`,
        ),
        void this.Qa(i.au, t, e, s)
      );
    this.u.k(o.C.du, new Date().valueOf());
    const n = t.device;
    n && n.os_version instanceof Promise
      ? n.os_version.then((s) => {
          (t.device.os_version = s), e();
        })
      : e();
  }
  li(t) {
    var e;
    null === (e = this.u) || void 0 === e || e.k(o.C.fu, t);
  }
  ui() {
    var t;
    return null === (t = this.u) || void 0 === t ? void 0 : t.j(o.C.fu);
  }
  Ru(t, e, s, i) {
    let r = this.ui();
    (null == r || isNaN(r)) && (r = e);
    const n = (t - i) / 1e3;
    return (r = Math.min(r + n / s, e)), r;
  }
  vu(t, e) {
    return Math.max(0, (1 - t) * e * 1e3);
  }
  Va() {
    var t, e, s;
    const i = { Za: !0, au: 0 },
      r = null === (t = this.u) || void 0 === t ? void 0 : t.j(o.C.du);
    if (null == r || isNaN(r)) return i;
    const n = (null === (e = this.yt) || void 0 === e ? void 0 : e.pu()) || -1,
      h = (null === (s = this.yt) || void 0 === s ? void 0 : s.gu()) || -1;
    if (-1 === n || -1 === h) return i;
    const a = new Date().valueOf();
    let u = this.Ru(a, n, h, r);
    return u < 1
      ? ((i.Za = !1), (i.au = this.vu(u, h)), i)
      : ((u = Math.trunc(u) - 1), this.li(u), i);
  }
  Xs() {
    this.Go.Xs();
  }
  Vs() {
    return this.Ko;
  }
  addSdkMetadata(t) {
    for (const e of t) -1 === this.Yh.indexOf(e) && this.Yh.push(e);
  }
}
