// @braze/web-sdk@5.2.0/src/models/server-config.js
import {
  GLOBAL_RATE_LIMIT_CAPACITY_DEFAULT as Zt,
  GLOBAL_RATE_LIMIT_REFILL_RATE_DEFAULT as hi,
} from "../common/constants.js";
export default class Gt {
  constructor(
    t = 0,
    i = [],
    s = [],
    h = [],
    e = null,
    l = null,
    r = { enabled: !1 },
    a = { enabled: !1, refresh_rate_limit: void 0 },
    n = { enabled: !0, capacity: Zt, refill_rate: hi },
  ) {
    (this.ol = t),
      (this.gl = i),
      (this.fl = s),
      (this.bl = h),
      (this.Cl = e),
      (this.cl = l),
      (this.ml = r),
      (this.mi = a),
      (this.al = n),
      (this.ol = t),
      (this.gl = i),
      (this.fl = s),
      (this.bl = h),
      (this.Cl = e),
      (this.cl = l),
      (this.ml = r),
      (this.mi = a),
      (this.al = n);
  }
  Y() {
    return {
      s: "5.2.0",
      l: this.ol,
      e: this.gl,
      a: this.fl,
      p: this.bl,
      m: this.Cl,
      v: this.cl,
      c: this.ml,
      f: this.mi,
      grl: this.al,
    };
  }
  static qn(t) {
    let i = t.l;
    return (
      "5.2.0" !== t.s && (i = 0),
      new Gt(i, t.e, t.a, t.p, t.m, t.v, t.c, t.f, t.grl)
    );
  }
}
