// @braze/web-sdk@5.2.0/src/InAppMessage/display/html-message-to-html.js
import X from "../../util/browser-detector.js";
import { logInAppMessageHtmlClick } from "../log-in-app-message-html-click.js";
import { getUser as ro } from "../../Core/get-user.js";
import { InAppMessage } from "../index.js";
import { keys as to } from "../../util/code-utils.js";
import { OperatingSystems as Y } from "../../util/device-constants.js";
import { parseQueryStringKeyValues as ut } from "../../util/url-utils.js";
import { WindowUtils as eo } from "../../util/window-utils.js";
import e from "../../managers/braze-instance.js";
import { logger as r } from "../../../shared-lib/index.js";
import { BRAZE_MUST_BE_INITIALIZED_ERROR as _ } from "../../common/constants.js";
import { nodeScriptReplace as ct } from "../utils/in-app-message-utils.js";
export default function at(t, o, n, s, i, u) {
  const c = document.createElement("iframe");
  c.setAttribute("title", "Modal Message"),
    i && (c.style.zIndex = (i + 1).toString());
  const a = (e) => {
      const o = e.getAttribute("href"),
        n = e.onclick;
      return (r) => {
        if (null != n && "function" == typeof n && !1 === n.bind(e)(r)) return;
        let i = ut(o).abButtonId;
        if (
          ((null != i && "" !== i) || (i = e.getAttribute("id") || void 0),
          null != o && "" !== o && 0 !== o.indexOf("#"))
        ) {
          const n =
              "blank" ===
              (e.getAttribute("target") || "").toLowerCase().replace("_", ""),
            u = s || t.openTarget === InAppMessage.OpenTarget.BLANK || n,
            a = () => {
              logInAppMessageHtmlClick(t, i, o), eo.openUri(o, u, r);
            };
          u ? a() : t.he(c, a);
        } else logInAppMessageHtmlClick(t, i, o || void 0);
        return r.stopPropagation(), !1;
      };
    },
    m = (t, e, o) => {
      const n = `([\\w]+)\\s*=\\s*document.createElement\\(['"]${o}['"]\\)`,
        r = t.match(new RegExp(n));
      if (r) {
        const o = `${r[1]}.setAttribute("nonce", "${e}")`;
        return `${t.slice(0, r.index + r[0].length)};${o};${t.slice(
          r.index + r[0].length,
        )}`;
      }
      return null;
    };
  if (
    ((c.onload = () => {
      let s = null;
      if (null != u) {
        (s = document.createElement("html")), (s.innerHTML = t.message || "");
        const e = s.getElementsByTagName("style");
        for (let t = 0; t < e.length; t++) e[t].setAttribute("nonce", u);
        const o = s.getElementsByTagName("script");
        for (let t = 0; t < o.length; t++) {
          o[t].setAttribute("nonce", u),
            (o[t].innerHTML = o[t].innerHTML.replace(
              /<style>/g,
              `<style nonce='${u}'>`,
            ));
          const e = m(o[t].innerHTML, u, "script");
          e && (o[t].innerHTML = e);
          const n = m(o[t].innerHTML, u, "style");
          n && (o[t].innerHTML = n);
        }
      }
      const i = c.contentWindow;
      i.focus();
      const l = i.document.querySelector("html");
      l && ((l.innerHTML = s ? s.innerHTML : t.message || ""), ct(l));
      const f = i.document.getElementsByTagName("head")[0];
      if (null != f) {
        if (t.be()) {
          const e = document.createElement("style");
          (e.innerHTML = t.css || ""),
            (e.id = t.ye() || ""),
            null != u && e.setAttribute("nonce", u),
            f.appendChild(e);
        }
        const e = i.document.createElement("base");
        null != e && (e.setAttribute("target", "_parent"), f.appendChild(e));
      }
      const d = i.document.getElementsByTagName("title");
      d && d.length > 0 && c.setAttribute("title", d[0].textContent || "");
      const p = {
          closeMessage: function () {
            t.he(c);
          },
          logClick: function () {
            logInAppMessageHtmlClick(t, ...arguments);
          },
          display: {},
          web: {},
        },
        requestPushPermission = function () {
          return function () {
            const t = arguments;
            import("../../Push/request-push-permission.js").then((o) => {
              e.so()
                ? o.requestPushPermission(...Array.prototype.slice.call(t))
                : r.error(_);
            });
          };
        },
        g = {
          requestImmediateDataFlush: function () {
            const t = arguments;
            import("../../Core/request-immediate-data-flush.js").then(
              ({ requestImmediateDataFlush: requestImmediateDataFlush }) => {
                e.so()
                  ? requestImmediateDataFlush(...Array.prototype.slice.call(t))
                  : r.error(_);
              },
            );
          },
          logCustomEvent: function () {
            const t = arguments;
            import("../../Core/log-custom-event.js").then(
              ({ logCustomEvent: logCustomEvent }) => {
                if (!e.so()) return void r.error(_);
                logCustomEvent(...Array.prototype.slice.call(t));
              },
            );
          },
          logPurchase: function () {
            const t = arguments;
            import("../../Core/log-purchase.js").then(
              ({ logPurchase: logPurchase }) => {
                if (!e.so()) return void r.error(_);
                logPurchase(...Array.prototype.slice.call(t));
              },
            );
          },
          unregisterPush: function () {
            const t = arguments;
            import("../../Push/unregister-push.js").then(
              ({ unregisterPush: unregisterPush }) => {
                e.so()
                  ? unregisterPush(...Array.prototype.slice.call(t))
                  : r.error(_);
              },
            );
          },
          requestPushPermission: requestPushPermission(),
          changeUser: function () {
            const t = arguments;
            import("../../Core/change-user.js").then(
              ({ changeUser: changeUser }) => {
                if (!e.so()) return void r.error(_);
                changeUser(...Array.prototype.slice.call(t));
              },
            );
          },
        },
        h = function (t) {
          return function () {
            g[t](...Array.prototype.slice.call(arguments));
          };
        };
      for (const t of to(g)) p[t] = h(t);
      const b = [
          "setFirstName",
          "setLastName",
          "setEmail",
          "setGender",
          "setDateOfBirth",
          "setCountry",
          "setHomeCity",
          "setEmailNotificationSubscriptionType",
          "setLanguage",
          "addAlias",
          "setPushNotificationSubscriptionType",
          "setPhoneNumber",
          "setCustomUserAttribute",
          "addToCustomAttributeArray",
          "removeFromCustomAttributeArray",
          "incrementCustomUserAttribute",
          "setCustomLocationAttribute",
          "addToSubscriptionGroup",
          "removeFromSubscriptionGroup",
        ],
        y = function (t) {
          return function () {
            const e = ro();
            e && e[t](...Array.prototype.slice.call(arguments));
          };
        },
        A = {};
      for (let t = 0; t < b.length; t++) A[b[t]] = y(b[t]);
      p.getUser = function () {
        return A;
      };
      const j = { showFeed: o },
        C = function (e) {
          return function () {
            const o = arguments;
            t.he(c, function () {
              j[e](...Array.prototype.slice.call(o));
            });
          };
        },
        v = p.display;
      for (const t of to(j)) v[t] = C(t);
      const P = { registerAppboyPushMessages: requestPushPermission() },
        w = function (t) {
          return function () {
            P[t](...Array.prototype.slice.call(arguments));
          };
        },
        E = p.web;
      for (const t of to(P)) E[t] = w(t);
      if (
        ((i.appboyBridge = p), (i.brazeBridge = p), t.Ae !== InAppMessage.Ce.je)
      ) {
        const t = i.document.getElementsByTagName("a");
        for (let e = 0; e < t.length; e++) t[e].onclick = a(t[e]);
        const e = i.document.getElementsByTagName("button");
        for (let t = 0; t < e.length; t++) e[t].onclick = a(e[t]);
      }
      const $ = i.document.body;
      if (null != $) {
        t.ve() && ($.id = t.htmlId || "");
        const e = document.createElement("hidden");
        (e.onclick = p.closeMessage),
          (e.className = "ab-programmatic-close-button"),
          $.appendChild(e);
      }
      i.dispatchEvent(new CustomEvent("ab.BridgeReady")),
        -1 !== c.className.indexOf("ab-start-hidden") &&
          ((c.className = c.className.replace("ab-start-hidden", "")), n(c)),
        document.activeElement !== c && c.focus();
    }),
    (c.className =
      "ab-in-app-message ab-start-hidden ab-html-message ab-modal-interactions"),
    X.OS === Y.io)
  ) {
    const e = document.createElement("div");
    return (
      (e.className = "ab-ios-scroll-wrapper"), e.appendChild(c), (t.Pe = e), e
    );
  }
  return (t.Pe = c), c;
}
