// @braze/web-sdk@5.2.0/src/triggers/models/trigger-condition.js
import ts from "./custom-event-data.js";
import rs from "./custom-event-property-data.js";
import is from "./filter-set.js";
import ni from "./in-app-message-click-data.js";
import ns from "./purchase-data.js";
import hs from "./purchase-property-data.js";
import ur from "./push-click-data.js";
import tt from "./trigger-events.js";
import { logger as r } from "../../../shared-lib/index.js";
import { getErrorMessage as si } from "../../util/error-utils.js";
export default class ri {
  constructor(e, t) {
    (this.type = e), (this.data = t), (this.type = e), (this.data = t);
  }
  sc(e, t) {
    return ri.cc[this.type] === e && (null == this.data || this.data.Jl(t));
  }
  static Gu(e, t) {
    let s = null;
    try {
      s = window.atob(e);
    } catch (t) {
      return r.info("Failed to unencode analytics id " + e + ": " + si(t)), !1;
    }
    return t === s.split("_")[0];
  }
  static fromJson(e) {
    const t = e.type;
    let r = null;
    switch (t) {
      case ri.Or.OPEN:
      case ri.Or.kt:
        break;
      case ri.Or.Rr:
        r = ns.fromJson(e.data);
        break;
      case ri.Or.nc:
        r = hs.fromJson(e.data);
        break;
      case ri.Or.vr:
        r = ur.fromJson(e.data);
        break;
      case ri.Or.ue:
        r = ts.fromJson(e.data);
        break;
      case ri.Or.lc:
        r = rs.fromJson(e.data);
        break;
      case ri.Or.Zr:
        r = ni.fromJson(e.data);
    }
    return new ri(t, r);
  }
  Y() {
    return { t: this.type, d: this.data ? this.data.Y() : null };
  }
  static qn(e) {
    let t,
      r = null;
    switch (e.t) {
      case ri.Or.OPEN:
      case ri.Or.kt:
        break;
      case ri.Or.Rr:
        r = new ns(e.d);
        break;
      case ri.Or.nc:
        (t = e.d || {}), (r = new hs(t.id, is.qn(t.pf || [])));
        break;
      case ri.Or.vr:
        r = new ur(e.d);
        break;
      case ri.Or.ue:
        r = new ts(e.d);
        break;
      case ri.Or.lc:
        (t = e.d || {}), (r = new rs(t.e, is.qn(t.pf || [])));
        break;
      case ri.Or.Zr:
        r = new ni(e.d);
    }
    return new ri(e.t, r);
  }
}
(ri.Or = {
  OPEN: "open",
  Rr: "purchase",
  nc: "purchase_property",
  vr: "push_click",
  ue: "custom_event",
  lc: "custom_event_property",
  Zr: "iam_click",
  kt: "test",
}),
  (ri.cc = {}),
  (ri.cc[ri.Or.OPEN] = tt.OPEN),
  (ri.cc[ri.Or.Rr] = tt.Rr),
  (ri.cc[ri.Or.nc] = tt.Rr),
  (ri.cc[ri.Or.vr] = tt.vr),
  (ri.cc[ri.Or.ue] = tt.ue),
  (ri.cc[ri.Or.lc] = tt.ue),
  (ri.cc[ri.Or.Zr] = tt.Zr),
  (ri.cc[ri.Or.kt] = tt.kt);
